File: AlgortihmEOS/Graph/7.cpp
#include <iostream>
#include <vector>

using namespace std;

const char EMPTY = '.';
const char WALL = '#';
const char HUMAN = 'H';
const char CORONA = 'C';

void solveCoronaMaze(vector<vector<char>>& grid);

bool putWalls(vector<vector<char>>& grid);
void dfs(vector<vector<char>>& grid, int i, int j);

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<char>> grid(n, vector<char>(m));

    for(int i = 0; i < n; i++) {
        string s;
        cin >> s;
        for(int j = 0; j < m; j++) {
            grid[i][j] = s[j];
        }
    }

    solveCoronaMaze(grid);
}

void solveCoronaMaze(vector<vector<char>>& grid) {
    int n = grid.size();
    int m = grid[0].size();

    if (putWalls(grid)) {
        bool flag = true;
        dfs(grid, n-1, m-1);
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if (grid[i][j] == HUMAN) {
                    flag = false;
                    break;
                }
            }
            if (!flag) break;
        }
        cout << (flag ? "YES" : "NO") << endl;
    } else {
        cout << "NO" << endl;   
    }
}

bool putWalls(vector<vector<char>>& grid) {
    vector<vector<int>> dirs = {
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1}
    };

    int n = grid.size();
    int m = grid[0].size();

    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            char &cur = grid[i][j];
            for(auto dir : dirs) {
                int x = i + dir[0];
                int y = j + dir[1];

                if (x >= 0 && y >= 0 && x < n && y < m) {
                    char &adj = grid[x][y];
                    if (cur == CORONA && adj == EMPTY) adj = WALL;
                    if (cur == CORONA && adj == HUMAN) return false;
                }
            }
        }
    }
    return true;
}

void dfs(vector<vector<char>>& grid, int i, int j) {
    vector<vector<int>> dirs = {
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1}
    };

    int n = grid.size();
    int m = grid[0].size();
    
    char &cur = grid[i][j];

    if (cur == WALL) return;
    cur = WALL;

    for(auto dir : dirs) {
        int x = i + dir[0];
        int y = j + dir[1];

        if (x >= 0 && y >= 0 && x < n && y < m) {
            dfs(grid, x, y);
        }
    }
}

File: AlgortihmEOS/Graph/7.py
n, m = map(int, input().split())
grid = [list(input()) for _ in range(n)]

EMPTY = '.'
WALL = '#'
HUMAN = 'H'
CORONA = 'C'

# Utility function to validate the cell
def is_valid(i, j):
    return 0 <= i < n and 0 <= j < m

# Utility function to get adjacent cells
def get_adj(i, j):
    adj = []
    
    dirs = [
        (0, 1),
        (0, -1),
        (1, 0),
        (-1, 0)
    ]

    for x, y in dirs:
        if is_valid(i+x, j+y):
            adj.append((i+x, j+y))

    return adj

def put_walls(grid):
    for i in range(n):
        for j in range(m):
            if grid[i][j] == CORONA:
                for x, y in get_adj(i, j):
                    if grid[x][y] == HUMAN:
                        return False
                    elif grid[x][y] == EMPTY:
                        grid[x][y] = WALL
    return True

def dfs(grid, i, j):
    if grid[i][j] == WALL: return
    grid[i][j] = WALL

    for x, y in get_adj(i, j):
        dfs(grid, x, y)

def solve_corona_maze(grid):
    if not put_walls(grid):
        print('NO')
    else:
        flag = True
        dfs(grid, n-1, m-1)

        for i in range(n):
            for j in range(m):
                if grid[i][j] == HUMAN:
                    flag = False
                    break
            if not flag:
                break

        print('YES' if flag else 'NO')
            


solve_corona_maze(grid)

